src/main/scala/org/allenai/plugins/DeployPlugin.scala

```diff
@@ -459,92 +486,306 @@ object DeployPlugin extends AutoPlugin {
+  }
+
+  /** Run `rsync` to copy the staged contents of this project to a remote host.
+    * @param host the remote host to `rsync` to
+    * @param sshCommand the SSH command to pass as the `rsh` option when running `rsync`
+    * @param srcDir the path to the local directory containing the staged contents of this project
+    * @param targetDir the path to the directory on the remote host to sync with
+    * @param deployedDirs the names of subdirectories of this project to copy to the remote host
+    */
+  def rsyncToRemote(
+    host: String,
+    sshCommand: Seq[String],
+    srcDir: String,
+    targetDir: String,
+    deployedDirs: Seq[String],
+    log: DeployLogger
+  ): Unit = {
+    // Build the `rsync` command.
+    val rsyncIncludes = deployedDirs map { name => s"--include=/$name" }
+    val rsyncCommand = Seq.concat(
+      Seq(
+        "rsync",
+        "-vcrtzP",
+        s"--rsh=${sshCommand.mkString(" ")}"
+      ),
+      rsyncIncludes,
+      Seq(
+        "--exclude=/*",
+        "--delete",
+        srcDir,
+        s"$host:$targetDir"
+      )
+    )
+    // Also build a shell-friendly version of the command to log out, for copy-paste.
+    val quotedRsync = rsyncCommand.updated(
+      2,
+      s"--rsh=${sshCommand.mkString("'", " ", "'")}"
+    ).mkString(" ")
+
+    // Log and run the command, and throw an error if it fails.
+    log.info(s"Running $quotedRsync . . .")
+    if (Process(rsyncCommand).! != 0) {
+      sys.error(s"Error while running rsync to host '$host'")
+    }
+  }
+
+  /** Write a Typesafe config object to file on a remote host without creating a local temp file.
+    * This is done by `echo`-ing the contents of the config locally, and piping the output into
+    * an SSH command running `cat`.
+    * We don't create a local temp file to avoid munging config overrides as many replicas deploy
+    * in parallel.
+    * The written file will be named `env.conf`.
+    * @param host the remote host to write the conf go
+    * @param envConf the Config object to write
+    * @param confPath the absolute path to the remote directory in which the written file should
+    *                 be placed
+    * @param sshCommand a constructed SSH command that should be used to communicate with the
+    *                   remote host
+    */
+  def copyEnvConfToRemote(
+    host: String,
+    envConf: Config,
+    confPath: String,
+    sshCommand: Seq[String],
+    log: DeployLogger
+  ): Unit = {
+    // Get the config contents to write.
+    val renderedConf = envConf.root().render(ConfigRenderOptions.concise)
+
+    // Command to print the full env config, to be piped to SSH.
+    val echoConfCommand = Seq("echo", renderedConf)
+    // Shell-friendly version of above, for copy-paste.
+    val quotedEcho = Seq("echo", s"'$renderedConf'").mkString(" ")
+
+    // SSH command to receive `echo` results and redirect them to disk.
+    val writeConfCommand = sshCommand :+ host :+ s"cat > $confPath/env.conf"
```

Comment:
This method is going to cause problems debugging, since there's no way to generate the `env.conf` file short of running the deploy.

Ideally, generating the `env.conf` file would be a separate SBT task, run before the actual deploy. That way, you could manually generate it (and you could just use `rsync` to push out the new file).